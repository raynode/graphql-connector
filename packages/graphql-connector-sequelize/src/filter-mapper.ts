
import {
  FilterMapper,
  isGraphQLBoolean,
  isGraphQLFloat,
  isGraphQLID,
  isGraphQLInt,
  isGraphQLString,
} from '@raynode/graphql-connector'
import {
  GraphQLBoolean,
  GraphQLInt,
  isListType,
  isScalarType,
  isEnumType,
} from 'graphql'

export const filterMapper: FilterMapper = (name, type, list, isFilter) => {
  if (isScalarType(type)) {
    if (isGraphQLID(type)) return {
        [`${name}`]: { type },
        [`${name}_in`]: { type: list },
      }
    if(isGraphQLString(type)) return {
        [`${name}`]: { type },
        [`${name}_in`]: { type: list },
        [`${name}_like`]: { type: list },
        [`${name}_startsWith`]: { type: list },
        [`${name}_endsWith`]: { type: list },
      }
    if (isGraphQLBoolean(type)) return { [`${name}`]: { type } }
    if (isGraphQLFloat(type) || isGraphQLInt(type))
      return {
        [`${name}`]: { type },
        [`${name}_gt`]: { type },
        [`${name}_lt`]: { type },
      }
  }
  if(isEnumType(type)) return {
    [`${name}`]: { type },
    [`${name}_in`]: { type: list },
  }
  if (isListType(type))
    return {
      [`${name}_contains`]: { type },
      [`${name}_hasLength`]: { type },
    }
  if (isFilter)
    return {
      [`${name}_empty`]: { type },
      [`${name}_some`]: { type },
    }
  return {}
}
