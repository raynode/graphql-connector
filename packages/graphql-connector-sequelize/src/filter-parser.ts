
import {
  FilterMapperMode,
  FilterParser,
} from '@raynode/graphql-connector'

import { Models } from './model-mapper'
import { DataTypes } from './type-guards'

import { Op } from 'sequelize'

// handle operations that sequelize does not support directly or update them to be more useful
const handlers = {
  startsWith: value => ({ [Op.like]: `%${value}` }),
  endsWith: value => ({ [Op.like]: `${value}%` }),
  like: value => ({ [Op.like]: `%${value}%` }),
}
// conversion function to get the correct operator
const getOperator = (handler: string, value: any) => handlers.hasOwnProperty(handler)
  ? handlers[handler](value)
  : Op.hasOwnProperty(handler)
    ? { [Op[handler]]: value }
    : null

// people who do not like recursion are people who do not ...
const applyFilterParser = <Key extends keyof Models>(
  mode: FilterMapperMode,
  model: Models[Key],
  data: Record<string, any>,
) => Object.keys(data).reduce((filters, name) => filterParser(mode, model, name, data[name], filters), {})

export const filterParser: FilterParser<DataTypes, Models> = (mode, model, name, value, filters) => {
  // convert all where-filters into the correct sequelize groups
  if(mode === 'where') {
    if(name === 'not') return {
      ...filters,
      [name]: getOperator(name, applyFilterParser(mode, model, value)),
    }
    if(name === 'and' || name === 'or')  return {
      ...filters,
      [name]: getOperator(name, value.map(entry => applyFilterParser(mode, model, entry))),
    }
    const [attribute, handler = 'eq'] = name.split('_')
    const operator = getOperator(handler, value)
    if(!operator)
      throw new Error(`unkown filter for ${name}`)
    filters[attribute] = {
      ...filters[attribute],
      ...operator,
    }
    return filters
  }
  // @TODO create
  // @TODO update
  return filters
}
