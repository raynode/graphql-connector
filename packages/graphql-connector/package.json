{
  "name": "@raynode/graphql-connector",
  "version": "0.0.1",
  "description": "",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "scripts": {
    "dev": "npm run test:watch",
    "build": "rimraf lib && tsc --p ./tsconfig.prod.json",
    "test": "npm run test:coverage",
    "test-only": "jest",
    "test:watch": "jest --watch",
    "test:coverage": "jest --coverage"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/raynode/graphql-connector.git"
  },
  "husky": {
    "hooks": {
      "pre-commit": "lint-staged"
    }
  },
  "lint-staged": {
    "src/**/*.{ts,tsx}": [
      "tslint --fix",
      "prettier --write",
      "git add"
    ]
  },
  "files": [
    "lib/*"
  ],
  "keywords": [],
  "author": "",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/raynode/graphql-connector/issues"
  },
  "homepage": "https://github.com/raynode/graphql-connector/src/graphql-connector#readme",
  "dependencies": {
    "@types/graphql": "^0.13.4",
    "@types/inflection": "^1.5.28",
    "@types/lodash": "^4.14.116",
    "graphql": "^0.13.2",
    "inflection": "^1.12.0",
    "lodash": "^4.17.10"
  },
  "devDependencies": {
    "@types/jest": "^23.3.9",
    "husky": "^1.1.2",
    "jest": "^23.5.0",
    "lint-staged": "^7.3.0",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "sequelize-cli": "^4.1.1",
    "ts-jest": "^23.10.4",
    "ts-node": "^7.0.1",
    "tsconfig-paths": "^3.5.0",
    "tslint": "^5.11.0",
    "typescript": "^3.1.3"
  },
  "gitHead": "150af815b281bbde4114efc0a0eb5bc608f9cc2d"
}
