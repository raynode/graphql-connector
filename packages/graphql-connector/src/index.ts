export * from './base-schema-generator'
export * from './naming-strategy'
export * from './type-converter'
export * from './filter-field-generator'
export * from './model'
export * from './model-mapper'
export * from './utils'
export * from './configuration'
export * from './order-mapper'

