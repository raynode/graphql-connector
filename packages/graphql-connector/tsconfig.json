{
  "compilerOptions": {
    "baseUrl": "./src",
    "declaration": true,
    "lib": ["es2016", "dom", "es7", "esnext"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noImplicitAny": false,
    "outDir": "lib",
    "rootDir": "src",
    "sourceMap": true,
    "target": "esnext"
  },
  "include": [
    "./src/**/*.ts"
  ]
}
