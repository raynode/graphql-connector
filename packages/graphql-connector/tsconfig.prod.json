{
  "extends": "./tsconfig.json",
  "exclude": [
    "src/**/*.spec.ts",
    "src/tests"
  ],
  "include": [
    "src"
  ]
}
