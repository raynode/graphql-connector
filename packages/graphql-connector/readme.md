
Graphql Connector
=================

This library is designed to easily connect a existing database system to the graphql ecosystem.
It takes the types and models of the source system and maps them to GraphQL.

Usage
-----

The connector needs to know 2 things to work:
* a ModelMapper to convert a Model from the source system into the GraphQL Connector Model-Type
* a TypeMapper to convert a Datatype of the source system into a GraphQLType

To see it in action you can look at the code for this example in the [tests](https://github.com/raynode/graphql-connector/blob/master/src/tests/example.spec.ts).
This code will be kept up to date and be used as a test

@TODO: subscriptions are not in yet

First step, define a graphql-connector model mapper
```Typescript
import {
  createModelMapper,
} from 'graphql-connector'

// myModels is expected to be { user: DBUserModel, post: DBPostModel, comment: DBCommentModel }
type Models = typeof myModels

// myTypes is expected to be { DBIntType: any, DBStringType: any, ... }
type Types = typeof myTypes

// utility to check if some type is a list-type
const isListType = (type: myTypes.DBType): type is myTypes.DBListType => type instanceof myTypes.DBListType

// utility to resolve the final type of a type
const getFinalType = (type: myTypes.DBType) => (isListType(type) ? type.subtype : type)

// The modelMapper is responsible to convert a my-model into a graphql-connector model
// a graphql-connector model needs attributes, associations
// each association and attribute could be setup with an resolver here
const modelMapper = createModelMapper<myTypes.DBType, Models>((model, addAttribute, addAssociation) => {
  // to convert a my-model into a graphql-connector model, we need to tell the system all attributes and associations
  Object.keys(model.attributes).forEach(name => {
    const attribute = model.attributes[name]
    const type = getFinalType(attribute)
    const list = isListType(attribute)
    if (name !== 'id' && type instanceof myTypes.DBIDType)
      return addAssociation({
        list,
        model: type.source as any,
        name,
      })
    return addAttribute({
      list,
      name,
      type,
    })
  })
})
```

Next step is to create a type mapper to convert our my-types into GraphQLType
```Typescript
import { TypeMapper } from 'graphql-connector'
// this function tells the generator how to convert my-type into a GraphQLType
const typeMapper: TypeMapper<Types, Models> = ({ type }) => {
  if (type instanceof myTypes.DBIDType) return GraphQLID
  if (type instanceof myTypes.DBDateType) return DateType
  if (type instanceof myTypes.DBIntType) return GraphQLInt
  if (type instanceof myTypes.DBStringType) return GraphQLString
  throw new Error('invalid type: ' + typeof type)
}
```

The last step is to input the typeMapper and the modelMapper into the createBaseSchema function and use the result to generate a BaseSchema
```Typescript
// Generate a baseSchemaGenerator by providing the modelMapper and typeMapper
// here we could change other things, like the nameingStrategy
const baseSchemaGenerator = createBaseSchemaGenerator({
  modelMapper,
  typeMapper,
})

// apply all my-models to the generator
// the baseSchema will have all queryFields, mutationsFields and subscriptionFields
const baseSchema = baseSchemaGenerator(myModels)
```

Now we have generated all required queryFields, mutationFields and subscriptionFields.
These now need to be converted to a GraphQLSchema, there is a utility function for it, but the step uses graphql code only.
```Typescript
// import { createSchema } from 'graphql-connector'
import {
  GraphQLObjectType,
  GraphQLSchema,
} from 'graphql'

export const createSchema = ({ queryFields, mutationFields }: BaseSchema) =>
  new GraphQLSchema({
    query: new GraphQLObjectType({
      name: 'Query',
      fields: queryFields,
    }),
    mutation: new GraphQLObjectType({
      name: 'Mutation',
      fields: mutationFields,
    }),
  })

const schema = createSchema(baseSchema)
```

This results in an SDL like:
```Graphql
type Comment implements Node {
  msg: String
  id: ID
  createdAt: Date
  lastUpdate: Date
  commentor: User
  Post: Post
}

type Comments implements List {
  nodes: [Comment!]!
  page: Page!
}

"""
A special custom Scalar type for Dates that converts to a ISO formatted string
"""
scalar Date

interface List {
  nodes: [Node!]!
  page: Page!
}

type Mutation {
  createUser: User
  updateUser: User
  deleteUsers: Users
  createPost: Post
  updatePost: Post
  deletePosts: Posts
  createComment: Comment
  updateComment: Comment
  deleteComments: Comments
}

interface Node {
  id: ID
}

type Page {
  limit: Int
  offset: Int
}

type Post implements Node {
  title: String
  text: String
  upvotes: Int
  id: ID
  createdAt: Date
  lastUpdate: Date
  author: User
  Comment: Comments
}

type Posts implements List {
  nodes: [Post!]!
  page: Page!
}

type Query {
  User: User
  Users: Users
  Post: Post
  Posts: Posts
  Comment: Comment
  Comments: Comments
}

type User implements Node {
  name: String
  email: String
  password: String
  id: ID
  createdAt: Date
  lastUpdate: Date
  Post: Posts
  Comment: Comments
}

type Users implements List {
  nodes: [User!]!
  page: Page!
}
```
