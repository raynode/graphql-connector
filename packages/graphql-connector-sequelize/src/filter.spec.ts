
import {
  GraphQLBoolean,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLString,
  GraphQLType,
} from 'graphql'
import { applyFilterParser } from '@raynode/graphql-connector'
import { filterMapper } from './filter-mapper'
import { filterParser } from './filter-parser'

const parser = applyFilterParser(filterParser)({} as any)
const map = (name: string, type: GraphQLType) => filterMapper(name, type, GraphQLList(GraphQLNonNull(type)), false)
const expectFilterMapper = (name: string, type: GraphQLType) => expect(map(name, type)).toMatchSnapshot(name)
const expectFilterParser = (where: Record<string, any>) => {
  const val = parser('where', where)
  expect(val).toMatchSnapshot()
}

describe('filter', () => {
  describe('mapping', () => {
    it('should create all filter mappings for basic types', () => {
      expectFilterMapper('int', GraphQLInt)
      expectFilterMapper('float', GraphQLFloat)
      expectFilterMapper('string', GraphQLString)
      expectFilterMapper('boolean', GraphQLBoolean)
      expectFilterMapper('id', GraphQLID)
    })
  })

  describe('parsing', () => {
    it('should parse correctly', () => {
      // should handle two together
      expectFilterParser({
        int_gt: 1,
        int_lt: 10,
      })
      // equality filter
      expectFilterParser({
        int: 4,
      })
      // all ID filters
      expectFilterParser({ id_in: [1, 2, 3] })
      expectFilterParser({ id: 'test' })
      // all string filters
      expectFilterParser({ string: 'test' })
      expectFilterParser({ string_in: ['a', 'b'] })
      expectFilterParser({ string_like: ['at'] }) // matches 'bat' and 'atmosphere'
      expectFilterParser({ string_startsWith: ['at'] }) // matches 'atmosphere'
      expectFilterParser({ string_endsWith: ['at'] }) // matches 'bat'
      // all boolean filters
      expectFilterParser({ boolean: true }) // matches 'bat'

      expectFilterParser({ not: {
        int_gt: 1,
        int_lt: 10,
      }})

      expectFilterParser({ and: [{ int_gt: 1 },  { int_lt: 10 }]})
      expectFilterParser({ or: [{ int_gt: 1 },  { int_lt: 10 }]})

      console.log('shift')
      console.log('shift')
      console.log('shift')
    })
  })
})
