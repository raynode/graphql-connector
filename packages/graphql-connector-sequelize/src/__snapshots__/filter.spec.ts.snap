// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter mapping should create all filter mappings for basic types: boolean 1`] = `
Object {
  "boolean": Object {
    "type": "Boolean",
  },
}
`;

exports[`filter mapping should create all filter mappings for basic types: float 1`] = `
Object {
  "float": Object {
    "type": "Float",
  },
  "float_gt": Object {
    "type": "Float",
  },
  "float_lt": Object {
    "type": "Float",
  },
}
`;

exports[`filter mapping should create all filter mappings for basic types: id 1`] = `
Object {
  "id": Object {
    "type": "ID",
  },
  "id_in": Object {
    "type": "[ID!]",
  },
}
`;

exports[`filter mapping should create all filter mappings for basic types: int 1`] = `
Object {
  "int": Object {
    "type": "Int",
  },
  "int_gt": Object {
    "type": "Int",
  },
  "int_lt": Object {
    "type": "Int",
  },
}
`;

exports[`filter mapping should create all filter mappings for basic types: string 1`] = `
Object {
  "string": Object {
    "type": "String",
  },
  "string_endsWith": Object {
    "type": "[String!]",
  },
  "string_in": Object {
    "type": "[String!]",
  },
  "string_like": Object {
    "type": "[String!]",
  },
  "string_startsWith": Object {
    "type": "[String!]",
  },
}
`;

exports[`filter parsing should parse correctly 1`] = `
Object {
  "int": Object {
    Symbol(gt): 1,
    Symbol(lt): 10,
  },
}
`;

exports[`filter parsing should parse correctly 2`] = `
Object {
  "int": Object {
    Symbol(eq): 4,
  },
}
`;

exports[`filter parsing should parse correctly 3`] = `
Object {
  "id": Object {
    Symbol(in): Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`filter parsing should parse correctly 4`] = `
Object {
  "id": Object {
    Symbol(eq): "test",
  },
}
`;

exports[`filter parsing should parse correctly 5`] = `
Object {
  "string": Object {
    Symbol(eq): "test",
  },
}
`;

exports[`filter parsing should parse correctly 6`] = `
Object {
  "string": Object {
    Symbol(in): Array [
      "a",
      "b",
    ],
  },
}
`;

exports[`filter parsing should parse correctly 7`] = `
Object {
  "string": Object {
    Symbol(like): "%at%",
  },
}
`;

exports[`filter parsing should parse correctly 8`] = `
Object {
  "string": Object {
    Symbol(like): "%at",
  },
}
`;

exports[`filter parsing should parse correctly 9`] = `
Object {
  "string": Object {
    Symbol(like): "at%",
  },
}
`;

exports[`filter parsing should parse correctly 10`] = `
Object {
  "boolean": Object {
    Symbol(eq): true,
  },
}
`;

exports[`filter parsing should parse correctly 11`] = `
Object {
  "not": Object {
    Symbol(not): Object {
      "int": Object {
        Symbol(gt): 1,
        Symbol(lt): 10,
      },
    },
  },
}
`;

exports[`filter parsing should parse correctly 12`] = `
Object {
  "and": Object {
    Symbol(and): Array [
      Object {
        "int": Object {
          Symbol(gt): 1,
        },
      },
      Object {
        "int": Object {
          Symbol(lt): 10,
        },
      },
    ],
  },
}
`;

exports[`filter parsing should parse correctly 13`] = `
Object {
  "or": Object {
    Symbol(or): Array [
      Object {
        "int": Object {
          Symbol(gt): 1,
        },
      },
      Object {
        "int": Object {
          Symbol(lt): 10,
        },
      },
    ],
  },
}
`;
